feat(mailer): introduce mailer component

innocenzi opened the pull request:
Closes https://github.com/tempestphp/tempest-framework/issues/149

This pull request adds a mailer component, as a wrapper around Symfony Mailer. The API is not polished, but most of the features are already there, testing included.

```php
return new SmtpMailerConfig(
    scheme: Scheme::SMTP,
    host: env('MAILER_SMTP_HOST', default: '127.0.0.0'),
    port: env('MAILER_SMTP_PORT', default: 2525),
    username: env('MAILER_SMTP_USERNAME', default: ''),
    password: env('MAILER_SMTP_PASSWORD', default: ''),
);
```

```php
$this->mailer->send(new ResetPasswordEmail($user, $resetUrl));

// or

$this->mailer->send(new GenericEmail(
    subject: 'Reset your password',
    to: $user->emailAddress,  [
    html: view('./reset-password.view.php', /* ... */),
));

// or

$this->mailer->send(
    GenericEmail::build()
        ->subject('Reset your password')
        ->to($this->emailAddress)
        ->html(view('./reset-password.view.php')), /* ... */
);
```

```php
final class ResetPasswordEmail implements Email
{
    public Envelope $envelope {
        get => new Envelope(
            subject: "Reset your password",
            to: $this->user->emailAddress,
        );
    }

    public Content $content {
        get => new Content(
            html: view('./reset-password.view.php', user: $this->user, url: $this->resetUrl),
        );
    }

    public function __construct(
        private readonly User $user,
        private readonly string $resetUrl,
    ) {}
}
```

By default, sending emails is disabled during tests. There's no risk of forgetting to fake the mailer and sending actual emails.

```php
$sent = $mailer->send(new ResetPasswordEmail($user, $resetUrl));

$sent->assertSee($user->fullName);
$sent->assertSee($resetUrl);
$sent->assertSentTo($user->emailAddress);
$sent->assertFrom('no-reply@tempestphp.com');
```

I only included SMTP and Postmark for now, but we can easily add more adapters.

innocenzi committed:
feat(mailer): add mailer package

innocenzi committed:
test: maybe fix tests for Windows

brendt commented:
I'll review next week, it's a busy weekend for me :)

innocenzi committed:
refactor: add `Address` and use underlying email for assertions

innocenzi committed:
feat: support headers

innocenzi committed:
feat: improve attachments

innocenzi committed:
style: apply fixes from qa

innocenzi committed:
test: hopefully fix tests for windows

innocenzi committed:
fix: add dependency on `tempest/support`

innocenzi committed:
feat: add round robin and ses transports

innocenzi commented:
I'm essentially done, but I still need opinions on the `Email` interfac, `GenericEmail` name and `EmailBuilder` names, which I'm not convinced about.

innocenzi committed:
Merge branch 'main' into feat/mailer

innocenzi committed:
chore: update to latest main

innocenzi committed:
refactor: update exceptions to follow conventions

innocenzi committed:
refactor: remove support for tagged mailers

innocenzi committed:
style: apply fixes from phpstan

innocenzi committed:
test: fix tests that did not reset the view renderer

brendt commented on packages/mailer/src/Email.php:
```diff
@@ -0,0 +1,20 @@
+    /**
+     * The envelope of the email.
+     */
+    public Envelope $envelope {
```
Referring back to my previous question: I'd rather have the properties encapsulated by `Envelope` and `Content` directly on the email instead

brendt commented on packages/mailer/src/EmailToSymfonyEmailMapper.php:
```diff
@@ -0,0 +1,106 @@
+
+    public function canMap(mixed $from, mixed $to): bool
+    {
+        if ($from instanceof Email) {
```
You probably also want to make sure `$to` is of the correct type — if possible? In my experience, automapping must be as specific as possible to avoid collisions with other mappers

brendt commented on packages/mailer/src/Exceptions/ExpeditorWasMissing.php:
```diff
@@ -0,0 +1,13 @@
+{
+    public function __construct()
+    {
+        parent::__construct('No expeditor address provided.');
```
I don't know what expeditor means? Maybe the exception could elaborate? Maybe I'm overthinking though

brendt commented on packages/mailer/src/GenericEmail.php:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Represents a generic email.
+ */
+final class GenericEmail implements Email
```
If the plan is still to encourage people to make their own Email implementations (I think they should), then we should provide a trait `IsEmail` as well for convenience. Certainly if/when `Envelope` and `Content` get refactored out.

innocenzi replied:
Maybe we shouldn't even have that `GenericEmail`, and encourage either creating email classes or using the email builder

In both cases, an `IsEmail` trait might indeed come in handy

brendt replied:
> Maybe we shouldn't even have that GenericEmail, and encourage either creating email classes or using the email builder

I like that :) Let's do that!

brendt commented on packages/mailer/src/GenericMailer.php:
```diff
@@ -0,0 +1,51 @@
+            ->with(fn (Email $from) => new EmailToSymfonyEmailMapper($this->mailerConfig, $this->viewRenderer)->map($from, null))
+            ->do();
+
+        return new SentGenericEmail(
```
Wouldn't it make more sense to dispatch an `EmailSent` event instead of returning a new value object? That way we can hook into the event from any place.

innocenzi replied:
This might be a good idea, but that changes the testing implementation a bit... I'll explore that

innocenzi replied:
I explored it and that makes testing too hard. But I did add event support 👍

brendt commented on packages/mailer/src/MailerConfig.php:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * The transport class.
+     */
+    public string $transport {
```
Add `@param class-string<Transport>` as the type

brendt commented on packages/mailer/src/MailerConfig.php:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * The default address from which emails will be sent.
+     */
+    public null|string|Address $from {
```
Maybe rename to `defaultFrom`?

innocenzi replied:
Renamed to `defaultExpeditor` to be even clearer

brendt commented on packages/mailer/src/Priority.php:
```diff
@@ -0,0 +1,31 @@
+
+namespace Tempest\Mail;
+
+enum Priority: int
```
Rename to `MailPriority`

innocenzi replied:
Good catch

brendt commented on packages/mailer/src/Priority.php:
```diff
@@ -0,0 +1,31 @@
+enum Priority: int
+{
+    /**
+     * The highest priority.
```
We don't really need these docblocks imo

innocenzi replied:
Yeah, I got lazy. Oops. I updated them to actually be useful.

brendt commented on packages/mailer/src/Transports/Postmark/Scheme.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Tempest\Mail\Transports\Postmark;
+
+enum Scheme: string
```
Rename to `PostmarkScheme` to avoid collisions

brendt commented on packages/mailer/src/Transports/Ses/Scheme.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Tempest\Mail\Transports\Ses;
+
+enum Scheme: string
```
Rename to `SesScheme` to avoid collisions

brendt commented on packages/mailer/src/Transports/Smtp/Scheme.php:
```diff
@@ -0,0 +1,9 @@
+
+namespace Tempest\Mail\Transports\Smtp;
+
+enum Scheme: string
```
Rename to `SmtpScheme` to avoid collisions

innocenzi committed:
refactor: remove `with` prefixes from email builder methods

innocenzi committed:
refactor: improve docblocks and naming

innocenzi committed:
refactor: disambiguate scheme classes

innocenzi committed:
refactor: clarify config `$from` property

innocenzi committed:
fix: ensure email mapper maps to symfony emails only

innocenzi commented on packages/mailer/src/Content.php:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Represents the content of an {@see \Tempest\Mailer\Envelope}.
+ */
+final class Content
```
I felt like the interface (`Email`) was super heavy with all those properties. It's a lot.

Though for content specifically, it could be in the interface directly. The envelope could also be shared between emails, so a dedicated object still makes sense. What do you think?

innocenzi commented on packages/mailer/src/Attachments/StorageAttachment.php:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Represents an attachment that lives in the {@see Tempest\Storage\Storage}.
+ */
+final class StorageAttachment implements Attachment
```
Unless I'm misunderstanding something, the developer experience is way better with dedicated objects and methods:

```php
$email = new Email()
    ->to('michael.scott@dundermifflin.com')
    // this is nice
    ->attachFromFileystem(__DIR__ . '/paperstock.csv')
    // this is super cumbersome
    ->attach(new Attachment(
        resolve: fn () => Filesystem\read_file(__DIR__ . '/paperstock.csv')
    ))
    ->text('Gotcha!')
    ->make();
```

```php
final class SendWelcomeEmail implements Email
{
    public Envelope $envelope {
        get => new Envelope(
            subject: "Welcome {$this->fullName}",
            to: $this->address,
            from: 'no-reply@tempestphp.com',
        );
    }

    public Content $content {
        get => new Content(
            html: view(__DIR__ . '/welcome.view.php', fullName: $this->fullName),
            // this is cool
            attachments: StorageAttachment::fromPath($this->welcomeCardName, tag: 'welcome-cards')
            // this is not cool
            attachments: new Attachment(
                resolve: fn () => get(Storage::class, 'welcome-cards')->readStream($this->welcomeCardName)
            )
        );
    }

    public function __construct(
        private readonly string $address,
        private readonly string $fullName,
        private readonly string $welcomeCardName,
    ) {}
}
```

innocenzi replied:
The alternative is an attachment builder class.

```php
new AttachmentBuilder()
    ->fromStorage($path, $tag)
    ->make();
```

innocenzi commented on packages/mailer/src/Builder/Email.php:
```diff
@@ -0,0 +1,209 @@
+    /**
+     * Adds an attachment from the filesystem.
+     */
+    public function withFileAttachment(string $path, ?string $name = null, ?string $contentType = null): self
```
See conversation in first review

brendt commented on packages/mailer/src/Builder/Email.php:
```diff
@@ -0,0 +1,209 @@
+use Tempest\View\View;
+use UnitEnum;
+
+final class Email
```
For clarification, the main problem I have with it is when you need both `Email` classes in the same context. You actually already run into the issue within the builder class itself where you need to import `Email as EmailInterface`. 

This happened way too much in my Laravel days and I cannot stand it 😅

innocenzi replied:
Alright, you convinced me. :p

innocenzi committed:
refactor: rename expeditor to sender

innocenzi committed:
refactor: simplify attachments

innocenzi committed:
refactor: rename email builder

innocenzi committed:
refactor: clean up configuration and user api

innocenzi committed:
feat: add event support

innocenzi commented:
Okay, after your review, are are the changes:

- `Envelope` and `Content` are still used because they are useful to avoid the use of a trait. However, the `GenericEmail` class can have parameters in its constructor, which avoids property-juggling and allows for a clean user-facing API:

```php
$this->mailer->send(new GenericEmail(
    subject: 'Hello',
    to: 'jon@doe.co',
    from: 'no-reply@tempestphp.com',
    text: 'Hello Jon',
    attachments: [
        Attachment::fromClosure(fn () => 'hello!'),
    ],
));
```

- I added a `build` method to `GenericEmail` and renamed the `Builer\Email` class to `EmailBuilder`. The intended use is now to call `GenericEmail::build()`, which conceptually makes a ton of sense.

- I removed the dedicated `DataAttachment`, `StorageAttachment` and `FileAttachment` classes in favor of a `Attachment` value object. This object has static constructors for attachments coming from the storage or the filesystem, for convenience:

```php
Attachment::fromStorage($storage, 'file.txt');
Attachment::fromFilesystem('./file.txt');
```

This is a good compromise; the `Attachment` class itself knows nothing about the storage, and there is no longer any service location, the storage has to be provided.

innocenzi committed:
test: fix line endings on windows

brendt commented on packages/mailer/src/EmailBuilder.php:
```diff
@@ -0,0 +1,192 @@
+    public function attach(Attachment ...$attachments): self
+    {
+        foreach ($attachments as $attachment) {
+            if (! ($attachment instanceof Attachment)) {
```
This has been checked by PHP's type system: `Attachment ...$attachments`

brendt commented on packages/mailer/src/Content.php:
```diff
@@ -0,0 +1,26 @@
+    /**
+     * @param Attachment[] $attachments
+     */
+    public function __construct(
```
Still not convinced we need `Content`. I'm ok with Envelope, but I think we need three parameters on the interface:

```php
public Envelope $envelope;
public string|View $content;
public array $attachments;
```

We loose the ability to differentiate between a view path and raw content (what used to be `$html` or `$text`). My proposal is that `$content` always assumes there's a view. Either it's a view object or a path to a view file. 

If we _really_ want to, we could have some kind of `raw` option in the builder which wraps whatever raw content in an instance of `GenericView`. Personally, I don't think we need `raw` atm, and I would just skip it for now, as it can be added later without needing a breaking change

brendt commented on packages/mailer/src/GenericEmail.php:
```diff
@@ -0,0 +1,54 @@
+
+    public function __construct(
+        public ?string $subject,
+        public null|string|array|Address $to,
```
I'm not a fan of the constructor on `GenericEmail` now accepting standalone parameters and then wrapping them in `Envelope`. I'd say just have people provide `Envelope`. (Assuming `Content` goes away)

brendt commented on packages/mailer/src/Testing/TestingMailer.php:
```diff
@@ -0,0 +1,45 @@
+            ->with(fn (Email $from) => new EmailToSymfonyEmailMapper($this->mailerConfig, $this->viewRenderer)->map($from, null))
+            ->do();
+
+        $this->eventBus?->dispatch(new EmailSent($email));
```
I'm going to find it confusing that we have `EmailSent` and `SentEmail`, and both meaning something different. Can't we merge both of them in the `EmailSent` event? I noticed there's a testing implementation for `SentEmail`, but I figure those assertion methods can be moved to somewhere else.

brendt commented:
Gonna work on this a bit myself

brendt committed:
Merge branch 'main' into feat/mailer

brendt committed:
wip

brendt committed:
wip

brendt committed:
wip

brendt committed:
wip

brendt committed:
wip

brendt committed:
wip

brendt committed:
wip

brendt committed:
wip

brendt committed:
wip

brendt committed:
Remove builder

brendt committed:
Cleanup tests

brendt committed:
Cleanup tests

brendt committed:
Cleanup interfaces

brendt committed:
wip

brendt committed:
wip

brendt committed:
wip

brendt commented:
Decided to not go with the view version for now. Let's first see how this goes

brendt committed:
wip

brendt committed:
Merge branch 'main' into feat/mailer
